File: server/routes/bundling.js
/**
 * Bundling Framework (GAP-008)
 *
 * Implements the 8-condition check from GFM SOP 2024 §3.1 to determine
 * whether a product variation qualifies for BUNDLING track (reduced SOPs).
 *
 * 8 Bundling Conditions:
 *   1. Existing approved parent product
 *   2. Same product category/family
 *   3. No new risk type introduced
 *   4. Within approved notional limits
 *   5. Same booking entity
 *   6. Same jurisdiction set
 *   7. No new counterparty type
 *   8. Within existing operational capacity
 *
 * All 8 must pass for BUNDLING track; otherwise routes to FULL_NPA or NPA_LITE.
 */

const express = require('express');
const router = express.Router();
const db = require('../db');

// GET /api/bundling/:id/assess — Run 8-condition bundling assessment
router.get('/:id/assess', async (req, res) => {
    try {
        const [project] = await db.query('SELECT * FROM npa_projects WHERE id = ?', [req.params.id]);
        if (project.length === 0) return res.status(404).json({ error: 'Project not found' });
        const npa = project[0];

        // Look up parent product if specified
        const parentId = req.query.parent_id || null;
        let parent = null;
        if (parentId) {
            const [parentRows] = await db.query('SELECT * FROM npa_projects WHERE id = ?', [parentId]);
            if (parentRows.length > 0) parent = parentRows[0];
        }

        const conditions = [];
        let allPassed = true;

        // Condition 1: Existing approved parent
        const c1 = parent && parent.current_stage === 'LAUNCHED';
        conditions.push({ id: 1, name: 'Existing Approved Parent Product', passed: !!c1, detail: c1 ? `Parent: ${parent.title}` : 'No approved parent product found' });
        if (!c1) allPassed = false;

        // Condition 2: Same product category
        const c2 = parent && npa.product_category === parent.product_category;
        conditions.push({ id: 2, name: 'Same Product Category', passed: !!c2, detail: c2 ? `Category: ${npa.product_category}` : `Category mismatch: ${npa.product_category} vs ${parent?.product_category || 'N/A'}` });
        if (!c2) allPassed = false;

        // Condition 3: No new risk type
        const [npaRisks] = await db.query('SELECT DISTINCT check_layer FROM npa_risk_checks WHERE project_id = ?', [npa.id]);
        const [parentRisks] = parent ? await db.query('SELECT DISTINCT check_layer FROM npa_risk_checks WHERE project_id = ?', [parent.id]) : [[]];
        const parentLayers = new Set(parentRisks.map((r) => r.check_layer));
        const newRisks = npaRisks.filter(r => !parentLayers.has(r.check_layer));
        const c3 = newRisks.length === 0;
        conditions.push({ id: 3, name: 'No New Risk Type', passed: c3, detail: c3 ? 'No new risk layers' : `New risk layers: ${newRisks.map(r => r.check_layer).join(', ')}` });
        if (!c3) allPassed = false;

        // Condition 4: Within approved notional limits
        const parentNotional = parseFloat(parent?.notional_amount) || 0;
        const npaNotional = parseFloat(npa.notional_amount) || 0;
        const c4 = parent && npaNotional <= parentNotional * 1.2; // Allow 20% buffer
        conditions.push({ id: 4, name: 'Within Notional Limits', passed: !!c4, detail: c4 ? `${npaNotional} <= ${parentNotional * 1.2} (120% of parent)` : `Exceeds limit: ${npaNotional} > ${parentNotional * 1.2}` });
        if (!c4) allPassed = false;

        // Condition 5: Same booking entity
        const [npaJurisdictions] = await db.query('SELECT jurisdiction_code FROM npa_jurisdictions WHERE project_id = ?', [npa.id]);
        const [parentJurisdictions] = parent ? await db.query('SELECT jurisdiction_code FROM npa_jurisdictions WHERE project_id = ?', [parent.id]) : [[]];
        // Simplified: check if primary booking location matches
        const c5 = parent && npa.currency === parent.currency; // proxy for booking entity
        conditions.push({ id: 5, name: 'Same Booking Entity', passed: !!c5, detail: c5 ? `Currency: ${npa.currency}` : 'Booking entity mismatch' });
        if (!c5) allPassed = false;

        // Condition 6: Same jurisdiction set
        const npaJSet = new Set(npaJurisdictions.map((j) => j.jurisdiction_code));
        const parentJSet = new Set(parentJurisdictions.map((j) => j.jurisdiction_code));
        const newJurisdictions = [...npaJSet].filter(j => !parentJSet.has(j));
        const c6 = newJurisdictions.length === 0;
        conditions.push({ id: 6, name: 'Same Jurisdiction Set', passed: c6, detail: c6 ? 'No new jurisdictions' : `New jurisdictions: ${newJurisdictions.join(', ')}` });
        if (!c6) allPassed = false;

        // Condition 7: No new counterparty type (check via form data)
        const [npaCounterparty] = await db.query("SELECT field_value FROM npa_form_data WHERE project_id = ? AND field_key = 'counterparty_type'", [npa.id]);
        const [parentCounterparty] = parent ? await db.query("SELECT field_value FROM npa_form_data WHERE project_id = ? AND field_key = 'counterparty_type'", [parent.id]) : [[]];
        const c7 = !npaCounterparty.length || (parentCounterparty.length > 0 && npaCounterparty[0]?.field_value === parentCounterparty[0]?.field_value);
        conditions.push({ id: 7, name: 'No New Counterparty Type', passed: c7, detail: c7 ? 'Same counterparty type' : 'New counterparty type detected' });
        if (!c7) allPassed = false;

        // Condition 8: Within operational capacity (simplified: check if ops signoff was approved on parent)
        let c8 = false;
        if (parent) {
            const [opsSignoff] = await db.query(
                "SELECT status FROM npa_signoffs WHERE project_id = ? AND party = 'Operations' AND status = 'APPROVED'",
                [parent.id]
            );
            c8 = opsSignoff.length > 0;
        }
        conditions.push({ id: 8, name: 'Within Operational Capacity', passed: c8, detail: c8 ? 'Operations capacity confirmed via parent' : 'No ops capacity confirmation' });
        if (!c8) allPassed = false;

        // Determine recommended track
        const passCount = conditions.filter(c => c.passed).length;
        let recommendedTrack = 'FULL_NPA';
        if (allPassed) {
            recommendedTrack = 'BUNDLING';
        } else if (passCount >= 6) {
            recommendedTrack = 'NPA_LITE';
        }

        res.json({
            project_id: npa.id,
            parent_id: parentId,
            conditions,
            all_passed: allPassed,
            pass_count: passCount,
            recommended_track: recommendedTrack
        });
    } catch (err) {
        console.error('[BUNDLING ERROR]', err.message);
        const errorMsg = process.env.NODE_ENV === 'production' ? 'Internal server error' : err.message;
        res.status(500).json({ error: errorMsg });
    }
});

// POST /api/bundling/:id/apply — Apply bundling track to NPA
router.post('/:id/apply', async (req, res) => {
    const { actor_name, parent_id } = req.body;
    try {
        await db.query(
            `UPDATE npa_projects SET approval_track = 'BUNDLING', updated_at = NOW() WHERE id = ?`,
            [req.params.id]
        );

        await db.query(
            `INSERT INTO npa_audit_log (project_id, actor_name, action_type, action_details, is_agent_action)
             VALUES (?, ?, 'BUNDLING_APPLIED', ?, 0)`,
            [req.params.id, actor_name || 'SYSTEM', JSON.stringify({ parent_id })]
        );

        res.json({ status: 'BUNDLING_APPLIED' });
    } catch (err) {
        console.error('[BUNDLING ERROR]', err.message);
        const errorMsg = process.env.NODE_ENV === 'production' ? 'Internal server error' : err.message;
        res.status(500).json({ error: errorMsg });
    }
});

module.exports = router;

File: server/mcp-python/tools/bundling.py
"""
Bundling Framework Tools — 2 tools (GAP-008)
8-condition check from GFM SOP 2024 §3.1 to determine whether a product
variation qualifies for BUNDLING track (reduced SOPs).
"""
import json

from registry import ToolDefinition, ToolResult, registry
from db import execute, query


# ─── Tool 1: bundling_assess ─────────────────────────────────────

BUNDLING_ASSESS_SCHEMA = {
    "type": "object",
    "properties": {
        "child_id": {"type": "string", "description": "NPA project ID of the child (new variation)"},
        "parent_id": {"type": "string", "description": "NPA project ID of the approved parent product"},
    },
    "required": ["child_id", "parent_id"],
}


async def bundling_assess_handler(inp: dict) -> ToolResult:
    child_rows = await query("SELECT * FROM npa_projects WHERE id = %s", [inp["child_id"]])
    if not child_rows:
        return ToolResult(success=False, error="Child project not found")
    child = child_rows[0]

    parent_rows = await query("SELECT * FROM npa_projects WHERE id = %s", [inp["parent_id"]])
    if not parent_rows:
        return ToolResult(success=False, error="Parent project not found")
    parent = parent_rows[0]

    conditions = []
    all_passed = True

    # C1: Existing approved parent
    c1 = parent.get("current_stage") == "LAUNCHED"
    conditions.append({"id": 1, "name": "Existing Approved Parent Product", "passed": c1,
                       "detail": f"Parent: {parent['title']}" if c1 else "No approved parent product found"})
    if not c1:
        all_passed = False

    # C2: Same product category
    c2 = child.get("product_category") == parent.get("product_category")
    conditions.append({"id": 2, "name": "Same Product Category", "passed": c2,
                       "detail": f"Category: {child.get('product_category')}" if c2 else
                       f"Mismatch: {child.get('product_category')} vs {parent.get('product_category')}"})
    if not c2:
        all_passed = False

    # C3: No new risk type
    child_risks = await query("SELECT DISTINCT check_layer FROM npa_risk_checks WHERE project_id = %s", [child["id"]])
    parent_risks = await query("SELECT DISTINCT check_layer FROM npa_risk_checks WHERE project_id = %s", [parent["id"]])
    parent_layers = {r["check_layer"] for r in parent_risks}
    new_risks = [r for r in child_risks if r["check_layer"] not in parent_layers]
    c3 = len(new_risks) == 0
    conditions.append({"id": 3, "name": "No New Risk Type", "passed": c3,
                       "detail": "No new risk layers" if c3 else
                       f"New risk layers: {', '.join(r['check_layer'] for r in new_risks)}"})
    if not c3:
        all_passed = False

    # C4: Within approved notional limits (120% buffer)
    parent_notional = float(parent.get("notional_amount") or 0)
    child_notional = float(child.get("notional_amount") or 0)
    c4 = parent_notional > 0 and child_notional <= parent_notional * 1.2
    conditions.append({"id": 4, "name": "Within Notional Limits", "passed": c4,
                       "detail": f"{child_notional} <= {parent_notional * 1.2} (120%)" if c4 else
                       f"Exceeds: {child_notional} > {parent_notional * 1.2}"})
    if not c4:
        all_passed = False

    # C5: Same booking entity (proxy: same currency)
    c5 = child.get("currency") == parent.get("currency")
    conditions.append({"id": 5, "name": "Same Booking Entity", "passed": c5,
                       "detail": f"Currency: {child.get('currency')}" if c5 else "Booking entity mismatch"})
    if not c5:
        all_passed = False

    # C6: Same jurisdiction set
    child_j = await query("SELECT jurisdiction_code FROM npa_jurisdictions WHERE project_id = %s", [child["id"]])
    parent_j = await query("SELECT jurisdiction_code FROM npa_jurisdictions WHERE project_id = %s", [parent["id"]])
    parent_jset = {j["jurisdiction_code"] for j in parent_j}
    new_j = [j["jurisdiction_code"] for j in child_j if j["jurisdiction_code"] not in parent_jset]
    c6 = len(new_j) == 0
    conditions.append({"id": 6, "name": "Same Jurisdiction Set", "passed": c6,
                       "detail": "No new jurisdictions" if c6 else f"New: {', '.join(new_j)}"})
    if not c6:
        all_passed = False

    # C7: No new counterparty type
    child_cp = await query(
        "SELECT field_value FROM npa_form_data WHERE project_id = %s AND field_key = 'counterparty_type'",
        [child["id"]])
    parent_cp = await query(
        "SELECT field_value FROM npa_form_data WHERE project_id = %s AND field_key = 'counterparty_type'",
        [parent["id"]])
    c7 = (not child_cp) or (parent_cp and child_cp[0].get("field_value") == parent_cp[0].get("field_value"))
    conditions.append({"id": 7, "name": "No New Counterparty Type", "passed": bool(c7),
                       "detail": "Same counterparty type" if c7 else "New counterparty type detected"})
    if not c7:
        all_passed = False

    # C8: Within operational capacity (ops signoff on parent)
    ops_signoff = await query(
        "SELECT status FROM npa_signoffs WHERE project_id = %s AND party = 'Operations' AND status = 'APPROVED'",
        [parent["id"]])
    c8 = len(ops_signoff) > 0
    conditions.append({"id": 8, "name": "Within Operational Capacity", "passed": c8,
                       "detail": "Ops capacity confirmed via parent" if c8 else "No ops capacity confirmation"})
    if not c8:
        all_passed = False

    pass_count = sum(1 for c in conditions if c["passed"])
    if all_passed:
        recommended = "BUNDLING"
    elif pass_count >= 6:
        recommended = "NPA_LITE"
    else:
        recommended = "FULL_NPA"

    return ToolResult(success=True, data={
        "child_id": inp["child_id"],
        "parent_id": inp["parent_id"],
        "parent_title": parent.get("title"),
        "conditions": conditions,
        "all_passed": all_passed,
        "pass_count": pass_count,
        "recommended_track": recommended,
    })


# ─── Tool 2: bundling_apply ─────────────────────────────────────

BUNDLING_APPLY_SCHEMA = {
    "type": "object",
    "properties": {
        "child_id": {"type": "string", "description": "NPA project ID to set as BUNDLING"},
        "parent_id": {"type": "string", "description": "Parent product NPA ID"},
        "actor_name": {"type": "string", "description": "Who applied the bundling track"},
    },
    "required": ["child_id", "parent_id"],
}


async def bundling_apply_handler(inp: dict) -> ToolResult:
    await execute(
        "UPDATE npa_projects SET approval_track = 'BUNDLING', updated_at = NOW() WHERE id = %s",
        [inp["child_id"]],
    )
    await execute(
        """INSERT INTO npa_audit_log (project_id, actor_name, action_type, action_details, is_agent_action)
           VALUES (%s, %s, 'BUNDLING_APPLIED', %s, 0)""",
        [inp["child_id"], inp.get("actor_name", "SYSTEM"),
         json.dumps({"parent_id": inp["parent_id"]})],
    )
    return ToolResult(success=True, data={
        "child_id": inp["child_id"],
        "parent_id": inp["parent_id"],
        "status": "BUNDLING_APPLIED",
    })


# ── Register ──────────────────────────────────────────────────────
registry.register_all([
    ToolDefinition(name="bundling_assess", description="Run the 8-condition bundling assessment comparing a child NPA against its parent product. Determines if BUNDLING track is eligible.", category="bundling", input_schema=BUNDLING_ASSESS_SCHEMA, handler=bundling_assess_handler),
    ToolDefinition(name="bundling_apply", description="Apply the BUNDLING approval track to an NPA after passing the 8-condition check.", category="bundling", input_schema=BUNDLING_APPLY_SCHEMA, handler=bundling_apply_handler),
])
